soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java

```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
+						Stmt stmt = (Stmt) unit;
+						//+++ collect handler fields signature	
+						if (stmt.containsFieldRef() && stmt.getFieldRef().getType().toString().contains("android.os.Handler")) {
+							if(stmt.getUseBoxes().size() > 1 && handlerInner.get(stmt.getFieldRef().getField().getSignature()) == null)
+								handlerInner.put(stmt.getFieldRef().getField().getSignature(),stmt.getUseBoxes().get(1).getValue().getType().toString());
+							// use value as locator
+							handlerClass.put(sootClass.getName() + ":" + stmt.getDefBoxes().get(0).getValue(),stmt.getFieldRef().getField().getSignature());
+						}
+							
+						if(stmt.containsInvokeExpr() && stmt.getInvokeExpr().getMethodRef().getSignature().contains("android.os.Handler")) {
+							//Soot has a access$XXX method that returns private fields in outer class
+							if(stmt.getInvokeExpr().getMethodRef().getName().contains("access")) {
+								Body b = stmt.getInvokeExpr().getMethod().retrieveActiveBody();
+								for (Iterator<Unit> u = b.getUnits().iterator(); u.hasNext();) {
+									Stmt s = (Stmt) u.next();
+									//if the access method return a handler, collect its signature
+									if(s.containsFieldRef() && s.getFieldRef().getType().toString().contains("android.os.Handler"))
+									{
+										if(!stmt.getDefBoxes().isEmpty())
+											handlerClass.put(sootClass.getName() + ":" + stmt.getDefBoxes().get(0).getValue(),s.getFieldRef().getField().getSignature());
+									}
+								}
+							}
+						}													
+					}
+
+					body.validate();
+				}
+
+			}
 
+								
+		}
+		//instrument the outerclass
+		for(String sc : handlerClass.keySet())
+			if(sc!=null)
+				generateSendMessage(sc.split(":")[0],handlerClass,handlerInner);
+
+	}
+
+	public void generateSendMessage(String sc,Map<String,String> handlerClass,Map<String,String> handlerInner) {
+		
+		if(sc == null || handlerClass.isEmpty() || handlerInner.isEmpty())
+			return;
+		SootClass sootClass = Scene.v().getSootClass(sc);
+		List<SootMethod> methodCopyList = new ArrayList<>(sootClass.getMethods());
+		for (SootMethod sootMethod : methodCopyList) {
+			if (sootMethod.isConcrete()) {
+				final Body body = sootMethod.retrieveActiveBody();
+				final LocalGenerator lg = Scene.v().createLocalGenerator(body);
+				
+				for (Iterator<Unit> unitIter = body.getUnits().snapshotIterator(); unitIter.hasNext();) {
+					Unit unit = unitIter.next();
+					Stmt stmt = (Stmt) unit;
+					if (stmt.containsInvokeExpr()) {
+						SootMethod callee = stmt.getInvokeExpr().getMethod();
+						
+						// For sendMessage(), we directly call the respective handler.handleMessage()
+						if (callee == smMessengerSend || callee == Scene.v().grabMethod("<android.os.Handler: boolean sendMessage(android.os.Message)>")) {
+							//collect the value for sendMessage()
+							String hc = handlerClass.get(sootClass.getName() + ":" + stmt.getInvokeExpr().getUseBoxes().get(1).getValue().toString());
```

Comment:
As stated above, you should have a proper object that stores the Soot objects instead of reyling on string parsing.